### ISSUE TITLE
Importing aluminium chloride fails after switching to isoelectronic model

### ISSUE THREAD

**Issue by ricrogz** [rdkit/rdkit#7788]
Before switching to the isoelectronic model, the following code worked:
```python3
mb = """
  SMMXDraw08051113052D

  0  0  0     0  0            999 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 8 8 0 0 0
M  V30 BEGIN ATOM
M  V30 1 Cl 11.5252 -12.6451 0.0000 0
M  V30 2 Al 12.7063 -12.6451 0.0000 0
M  V30 3 Cl 12.7063 -11.4640 0.0000 0
M  V30 4 Cl 12.7063 -13.8979 0.0000 0
M  V30 5 Al 14.7407 -13.8979 0.0000 0
M  V30 6 Cl 14.7407 -12.6451 0.0000 0
M  V30 7 Cl 15.9218 -13.8979 0.0000 0
M  V30 8 Cl 14.7407 -15.0790 0.0000 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 2 3 
M  V30 2 1 5 6 
M  V30 3 1 1 2 
M  V30 4 1 5 7 
M  V30 5 1 2 4 
M  V30 6 1 5 8 
M  V30 7 9 6 2 
M  V30 8 9 4 5 
M  V30 END BOND
M  V30 END CTAB
M  END

$$$$
"""

m = Chem.MolFromMolBlock(mb)
m.Debug()
---------------------------------------------------------------------------
Atoms:
	0 17 Cl chg: 0  deg: 1 exp: 1 imp: 0 hyb: SP3
	1 13 Al chg: 0  deg: 4 exp: 4 imp: 2 hyb: SP3D
	2 17 Cl chg: 0  deg: 1 exp: 1 imp: 0 hyb: SP3
	3 17 Cl chg: 0  deg: 2 exp: 1 imp: 0 hyb: SP3
	4 13 Al chg: 0  deg: 4 exp: 4 imp: 2 hyb: SP3D
	5 17 Cl chg: 0  deg: 2 exp: 1 imp: 0 hyb: SP3
	6 17 Cl chg: 0  deg: 1 exp: 1 imp: 0 hyb: SP3
	7 17 Cl chg: 0  deg: 1 exp: 1 imp: 0 hyb: SP3
Bonds:
	0 1->2 order: 1
	1 4->5 order: 1
	2 0->1 order: 1
	3 4->6 order: 1
	4 1->3 order: 1
	5 4->7 order: 1
	6 5->1 order: D
	7 3->4 order: D

```

But it fails after PR #7491:
```python3
m = Chem.MolFromMolBlock(mb)
---------------------------------------------------------------------------
[18:29:35] Explicit valence for atom # 1 Al, 4, is greater than permitted


assert m
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 assert m 

AssertionError: 

```

**Comment by greglandrum** [comment]
This happens because we removed the allowed oxidation state of 6 from the Al atom. So the hypervalent AlCl3 dimer as is present in that mol file is not supported with normal sanitization.

If you look at the debug output from before you will see that the Al had two implicit Hs assigned; this is wrong.

**Comment by ricrogz** [comment]
I was actually hypothesizing about the valence contribution of the dative bond. Currently, we have the dative bond to not contribute to the origin atom's valence, but contribute 1 to the end atom's valence. That's the reason the Al atom has valence 4 now. With the behavior we had before, the code assumed 2 implicit Hs so that it would fit with valence 6.

My question is: is this behavior of the valence contribution of the dative bonds still correct when we consider the isoelectronic model?

**Comment by greglandrum** [comment]
I think so. That change was really just intended to make it easier to figure out what the valence should be.

If you have other ideas, we can definitely discuss

**Comment by ricrogz** [comment]
Well, like I said, I'm wondering if that approach is still correct, since the dative bond if "lending" a pair of electrons to the atom at the end of the bond. I wonder if, instead of having the dative bond contribute 1 to the valence, we should make it consider the receiving atom as the equivalent isoelectronic atom.

**Comment by greglandrum** [comment]
I agree (and @ptosco definitely agrees) that we need to have a discussion about how dative bonds are handled, but that's not directly connected to the removal of the valence state 6 for Al.

**Comment by ricrogz** [comment]
I agree that the old version adding the 2 H atoms to the Al was incorrect, but that and the valence 6 on Al allowed this molecule to be imported with sanitization. This has always been broken, but in some way, it worked.

After switching to the isoelectronic valence model and removing valence 6, this is completely broken.

First, because the electrons donated through the dative bond are never considered when calculating the "effective" atomic number. Not in the case of Al in the aluminium chloride, nor for any other element involved in a dative bond, which is fine for elements with "catch all" valences (-1), as transition metals, but will cause trouble for any atom that doesn't have it.

And second, because the contribution of the dative bond to the valence of the atom being 1 doesn't make sense either. I think the proper solution should be to consider the 2 donated electrons to contribute to the valence, or if we want to simplify things, ignore the pair and consider a contribution of 0 electrons to the valence (I suspect we'd need 2 if we want to be correct about the hybridizations).

This is relevant because we use `bnd->getValenceContrib(&atom)` in `Atom::calculateExplicitValence()`.

The current value of 1 brings the valence of Al in aluminium chloride to 4, which doesn't match Al's valence of 3. But, as you mentioned, before the switch, when adding 2 Hs, it matched valence 6.

With the isoelectronic valence model and considering the 2 electrons from the dative bond, the "effective" element for the Al would be P, which supports valences 3 and 5. If we consider the dative bond to contribute either 0 or 2 to the valence, that's exactly the values we would get for the valence, allowing Aluminium chloride to be sanitizable.

**Comment by ptosco** [comment]
@ricrogz As we discussed at the UGM, this is what the "dative bond" PR that I am working on deals with.

However, changing the value returned by `getValenceContrib()` for `Bond::DATIVE` (which is most likely the right thing to do) is a major breaking change, which Greg thinks would be better suited for 2025.03 which will feature other major changes related to how implicit/explicit valence is handled, and I agree with him.

The PR I have been working on is attempt to preserve backwards compatibility while fixing problems such as the one you reported. I'll try my best to submit it ASAP, but I have several other things on my plate, so I cannot guarantee I'll manage.

**Comment by ricrogz** [comment]
@ptosco, @greglandrum. I know. I'm just arguing that the switch to the isoelectronic valence model is, in my understanding, incomplete without these fixes, and also a breaking change. Aluminium chloride is just a simple example. Paolo had a bunch other examples. Maybe switching to the isoelectronic model is not a major breaking change (unless you frequently work with dative bonds), but I'm still concerned about releasing something that we know is incomplete and causes trouble for some people.

**Comment by greglandrum** [comment]
It definitely changes things, and is documented as such, but this is not related to dative bonds.

The reason your Al example is behaving differently is because we removed the chemically unreasonable valence state of 6 for aluminum
